New introduction for the Gradle BT manual (#24831)
gradle/gradle: subprojects/docs/src/docs/userguide/what_is_gradle.adoc

[[what_is_gradle]]
= What is Gradle?

> Gradle is a https://en.wikipedia.org/wiki/Build_automation[build automation] tool. It is available as an open-source Build Tool and an Enterprise tool.

Reviewer: Hmmm I'm not sure about the wording here - there is a strong separation of what Gradle Enterprise is and what Gradle Build Tool is. Reading this I get the impression they are the same product offered in either open-source or enterprise modes. However they are separate products that provide complementary features.

Gradle Enterprise works with other build tools too, its not tied only to Gradle Build Tool.

Lets get input from others too.

Author: I made further changes, please let me know what you think.

[[what_is_gradle]]
= What is Gradle?

Gradle is a https://en.wikipedia.org/wiki/Build_automation[build automation] tool.

== Gradle Build Tool

Gradle Build Tool is the fast and dependable open source build system that automates building software of any type, size or complexity.

== Gradle Enterprise

> https://gradle.com/[Gradle Enterprise] is a powerful subscription based tool for managing and monitoring complex software builds. It provides teams with centralized visibility into their build process to quickly identify and address issues as they arise.

Reviewer 2: We tend to use one sentence per line in `.adoc` files to make it easier to read diffs when making changes.

Author: Noted. Thank you for this tip!

* Android builds add many new build concepts such as flavors and build types.
IDE support::
* Several major IDEs provide interaction with Gradle builds, including Android Studio, IntelliJ IDEA, Eclipse, VSCode, and NetBeans.
* Gradle can also generate the solution files required to load a project into Visual Studio.
Insight::
* https://scans.gradle.com/[Build Scan™] provides extensive information about a build that you can use to identify issues.
* You can use Build Scans™ to identify problems with a build's performance and even share them for debugging help.

== User Manual

The Gradle Build Tool user manual is broken down into the following sections:

Quick Start :: Get up and running installing Gradle.
Running Gradle Builds :: Use Gradle Build Tool to build your project.
Authoring Gradle Builds :: Develop tasks and plugins to customize your build.
Authoring JVM Builds :: Use Gradle Build Tool with your Java project.
Working with Dependencies :: Add dependencies to your build.
Optimizing Build Times :: Use a cache to optimize your build.
Gradle on CI :: Gradle integration with popular continuous integration (CI) servers.

== Fundamentals

> Users provide instructions to Gradle using either Groovy or Kotlin in a script to build a Project (your code).

Reviewer 2: Let's put Kotlin first now that this is the default for new builds
```suggestion
Users provide instructions to Gradle using either Kotlin or Groovy in a script to build a Project (your code).
```

Author: I am committing this change.

[[what_is_gradle]]
= What is Gradle?

> Gradle is a https://en.wikipedia.org/wiki/Build_automation[build automation] tool.

Author: This has been updated in the latest commit, let me know what you think @akshayabd

> Gradle, Inc. is the software company behind Gradle Build Tool (www.gradle.org) and Gradle Enterprise (www.gradle.com).

Reviewer: Hmmm I would be ok with just removing this line and start with the Gradle Build Tool section. Not sure how others feel.

Author: I made this change based on a suggestion by [ljacomet](https://github.com/ljacomet). I do need something that ties GE and BT together somewhere so users understand this is the BT manual.

== Gradle Build Tool

Gradle Build Tool (typically referred to as Gradle in this User Manual) is the fast and dependable open source build system that automates building software of any type, size or complexity.

== Gradle Enterprise

https://gradle.com/[Gradle Enterprise] is the powerful subscription based tool for managing and monitoring complex software builds.
It provides teams with centralized visibility into their build process to quickly identify and address issues as they arise.

== Why Gradle?

Gradle makes building and running apps easy; it is great for both small and large projects.

Gradle is fast, highly configurable, and very concise (no XML needed).

> Gradle makes few assumptions about what you’re trying to build or how to build it which makes it particularly flexible.

Author: Fixed in latest commit. Let me know what you think.

> Gradle makes few assumptions about what you’re trying to build or how to build it which makes it particularly flexible.

Reviewer 3: I wonder about the complete removal. At the same time, the section is then closer to our vision statement.

Maybe with sentences like:
> Gradle supports conventions to reduce configuration.
> It also is flexible when your use case differs from those conventions.

which could be between "Gradle is fast ..." and "The unique advantage ..."

@pioterj wdyt?

> Gradle makes few assumptions about what you’re trying to build or how to build it which makes it particularly flexible.

Author: I like this and updated the language accordingly in the latest commit. Thank you @ljacomet.

> Gradle makes few assumptions about what you’re trying to build or how to build it which makes it particularly flexible.

Author: I updated this section in the latest commit and reviewed it with @pioterj in out 1:1

* But these conventions don't limit you: you can configure settings, add your own tasks, and make many other customizations in your builds.
Extensibility::
* Most builds have special requirements that require custom build logic.
* You can readily extend Gradle to provide your own build logic with custom tasks and plugins.
* Android builds add many new build concepts such as flavors and build types.
IDE support::
* Several major IDEs provide interaction with Gradle builds, including Android Studio, IntelliJ IDEA, Eclipse, VSCode, and NetBeans.
* Gradle can also generate the solution files required to load a project into Visual Studio.
Insight::
* https://scans.gradle.com/[Build Scan™] provides extensive information about a build that you can use to identify issues.
* You can use Build Scans™ to identify problems with a build's performance and even share them for debugging help.

== User Manual

The Gradle Build Tool user manual is broken down into the following sections:

Quick Start :: Get up and running installing Gradle.
Running Gradle Builds :: Use Gradle Build Tool to build your project.

> Authoring Gradle Builds :: Develop tasks and plugins to customize your build.

Author: For the time being I'm going to keep the existing menu / flow of the user manual. I agree that it needs to be updated but this will be done at a later time.

Projects are the things that Gradle builds.
Projects contain a build script, which is a file located in the project's root directory usually named `build.gradle` or `build.gradle.kts`.

Builds scripts define tasks, dependencies, plugins, and other configuration for that project.
A single build can contain one or more projects and each project can contain their own subprojects.

=== Builds

> A build is an execution of a collection of tasks in a Gradle project.

Reviewer 2: This contradicts the previous sentence:

> A single build can contain one or more projects and each project can contain their own subprojects.

What about the following?
```suggestion
A build invocation is an execution of a collection of tasks in the projects contained by a build.
```

Probably not the best sentence but at least the terminology is clearer to me. WDYT?

Author: I am thinking:
```suggestion
A build invocation is the execution of a collection of tasks in the project(s).
```
Thoughts?

Reviewer 2: Sounds good! :+1:

Author: Updated in the latest commit.

Projects are the things that Gradle builds.
Projects contain a build script, which is a file located in the project's root directory usually named `build.gradle` or `build.gradle.kts`.

Builds scripts define tasks, dependencies, plugins, and other configuration for that project.
A single build can contain one or more projects and each project can contain their own subprojects.

=== Builds

A build invocation is an execution of a collection of tasks in the project(s).
You run a build via the command line interface (CLI) or an IDE by specifying task selectors.

> Gradle configures the build and selects the tasks to run.

Reviewer: cc @big-guy

> Gradle configures the build and selects the tasks to run.

Reviewer 4: "configures the build" makes sense to me as part of the intro.  The words "build" and "configuration" have so many overloaded meanings, that keeping it familiar with the previous sentence ("You run a build...") is probably better than introducing the term "build configuration".
